File: src/app/model/SpeedMetric.ts
import { LocationServiceEvent } from './LocationServiceEvent';
import { MetricService } from "../services/metric.service";
import { Metric, MetricType } from "./Metric";
import { SpeedSensorServiceEvent } from './SpeedSensorServiceEvent';
import { SpeedSensorData } from './SpeedSensorData';
import { last } from 'rxjs';

export class SpeedMetric extends Metric {
    private locationHandler = this.locationEventHandler.bind(this);
    private speedSensorHandler = this.speedSensorEventHandler.bind(this);
    private lastSpeedSensorDatas: SpeedSensorData[] = [];

    constructor(metricService: MetricService, private wheelCircumference: number = 2.1) {
        super(MetricType.SPEED, 'Speed', 'm/s', metricService, 'km/h');
    }

    startLogging(): void {
        if (this.metricService.getSpeedSensorService().deviceSelected) {
            this.metricService.getSpeedSensorService().subscribeForSpeedData(this.speedSensorHandler);
        } else {
            this.metricService.getLocationService().subscribeForLocation(this.locationHandler);
        }
    }

    stopLogging(): void {
        this.metricService.getLocationService().unsubscribeForLocation(this.locationHandler);
        this.metricService.getSpeedSensorService().unsubscribeForSpeedData(this.speedSensorHandler);
    }

    locationEventHandler(event: LocationServiceEvent): void {
        const speed = event.location.coords.speed;
        if (speed !== null) {
            this.addValue(speed, new Date(event.location.timestamp));
        } else {
            // If speed is null, we can set it to 0 or handle it as needed
            this.addValue(0, new Date(event.location.timestamp));
        }
    }

    speedSensorEventHandler(event: SpeedSensorServiceEvent): void {
        if (!event.speedSensorData.cumulativeWheelRevolutions) {
            return;
        }

        const lastDataWithLessWheelRevolutions = this.getLastSpeedSensorDataWithLessWheelRevolutions(event.speedSensorData.cumulativeWheelRevolutions);
        if (lastDataWithLessWheelRevolutions === null) {
            this.addValue(0, new Date(event.speedSensorData.timestamp));
        } else {
            const diffRotations = event.speedSensorData.diffWheelRevolutions(lastDataWithLessWheelRevolutions) || 0;
            const timediff = event.speedSensorData.diffWheelTimeInSeconds(lastDataWithLessWheelRevolutions) || 1;
            const speed = (diffRotations * this.wheelCircumference) / timediff; // in m/s
            console.log(diffRotations, timediff, speed);
            this.addValue(speed, event.speedSensorData.timestamp);
        }
        this.lastSpeedSensorDatas.push(event.speedSensorData);
        if (this.lastSpeedSensorDatas.length > 10) {
            this.lastSpeedSensorDatas.shift();
        }
    }

    getLastSpeedSensorDataWithLessWheelRevolutions(cumulativeWheelRevolutions: number): SpeedSensorData | null {
        for (let i = this.lastSpeedSensorDatas.length - 1; i >= 0; i--) {
            const data = this.lastSpeedSensorDatas[i];
            if (data.cumulativeWheelRevolutions && data.cumulativeWheelRevolutions < cumulativeWheelRevolutions) {
                return data;
            }
        }
        return null;
    }

}

File: android/app/src/main/java/com/openbikecomputer/model/SpeedMetric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;
import com.openbikecomputer.service.OnLocationUpdate;
import com.openbikecomputer.service.OnSpeedData;

import java.util.ArrayList;
import java.util.List;

public class SpeedMetric extends Metric implements OnLocationUpdate, OnSpeedData {
    private final double wheelCircumference;
    private final List<SpeedSensorData> lastSpeedSensorDatas = new ArrayList<>();

    public SpeedMetric(IMetricService metricService) {
        super(MetricType.SPEED, "Speed", "m/s", metricService, "km/h", 1);
        this.wheelCircumference = 2.1;
    }

    @Override
    public void startLogging() {
        metricService.getSpeedSensorService().subscribeForSpeedData(this);
        metricService.getLocationService().subscribeForLocation(this);
    }

    @Override
    public void stopLogging() {
        metricService.getLocationService().unsubscribeForLocation(this);
        metricService.getSpeedSensorService().unsubscribeForSpeedData(this);
    }

    @Override
    public void onLocation(LocationSample sample) {
        if (!metricService.getSpeedSensorService().isDeviceSelected()) {
            if (sample.speed != null) {
                addValue(sample.speed, sample.timestamp);
            } else {
                addValue(0, sample.timestamp);
            }
        }
    }

    @Override
    public void onSpeedData(SpeedSensorData data) {
        if (data.cumulativeWheelRevolutions == null || data.cumulativeWheelRevolutions == 0) {
            return;
        }

        SpeedSensorData lastDataWithLessWheelRevolutions =
                getLastSpeedSensorDataWithLessWheelRevolutions(data.cumulativeWheelRevolutions);
        if (lastDataWithLessWheelRevolutions == null) {
            addValue(0, data.timestamp);
        } else {
            Double diffRotationsObj = data.diffWheelRevolutions(lastDataWithLessWheelRevolutions);
            Double timeDiffObj = data.diffWheelTimeInSeconds(lastDataWithLessWheelRevolutions);
            double diffRotations = diffRotationsObj == null ? 0 : diffRotationsObj;
            double timeDiff = timeDiffObj == null || timeDiffObj == 0 ? 1 : timeDiffObj;
            addValue((diffRotations * wheelCircumference) / timeDiff, data.timestamp);
        }
        lastSpeedSensorDatas.add(data);
        if (lastSpeedSensorDatas.size() > 10) {
            lastSpeedSensorDatas.remove(0);
        }
    }

    public SpeedSensorData getLastSpeedSensorDataWithLessWheelRevolutions(long cumulativeWheelRevolutions) {
        for (int i = lastSpeedSensorDatas.size() - 1; i >= 0; i--) {
            SpeedSensorData data = lastSpeedSensorDatas.get(i);
            if (data.cumulativeWheelRevolutions != null
                    && data.cumulativeWheelRevolutions != 0
                    && data.cumulativeWheelRevolutions < cumulativeWheelRevolutions) {
                return data;
            }
        }
        return null;
    }
}
